From: "J. R. Okajima" <hooanon05@yahoo.co.jp>
Date: Tue, 25 Jun 2013 17:42:44 +0900
Subject: [45/83] aufs: bugfix, retval from au_pin() call in
 au_file_refresh_by_inode()
Origin: http://sourceforge.net/p/aufs/aufs3-standalone/ci/6e7d81840dae19c0d774c14d0c652f5f2b00f98b

Simply it is a bug to call au_unpin() when the precedding au_pin()
returns an error.
To be honest, I was very surprised that this bug was born in aufs1 era
and lived very long time.

Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>
[bwh: Adjust context]
---
 fs/aufs/file.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/fs/aufs/file.c
+++ b/fs/aufs/file.c
@@ -422,10 +422,11 @@ static int au_file_refresh_by_inode(stru
 		/* always superio. */
 		err = au_pin(&pin, dentry, bstart, AuOpt_UDBA_NONE,
 			     AuPin_DI_LOCKED | AuPin_MNT_WRITE);
-		if (!err)
+		if (!err) {
 			err = au_sio_cpup_simple(dentry, bstart, -1,
 						 AuCpup_DTIME);
-		au_unpin(&pin);
+			au_unpin(&pin);
+		}
 	} else if (hi_wh) {
 		/* already copied-up after unlink */
 		err = au_reopen_wh(file, bstart, hi_wh);
